examples/wlr-workspace.c

```diff
@@ -0,0 +1,338 @@
+		struct workspace_v1 *workspace;
+		wl_list_for_each(workspace, &group->workspaces, link) {
+			copy_state(&workspace->current, &workspace->pending);
+			print_workspace(workspace);
+		}
+	}
+}
+
+static void workspace_manager_handle_finished(void *data,
+		struct zwlr_workspace_manager_v1 *zwlr_workspace_manager_v1) {
+	zwlr_workspace_manager_v1_destroy(zwlr_workspace_manager_v1);
+}
+
+static const struct zwlr_workspace_manager_v1_listener workspace_manager_impl = {
+	.workspace_group = workspace_manager_handle_workspace_group,
+	.done = workspace_manager_handle_done,
+	.finished = workspace_manager_handle_finished,
+};
+
+struct wl_seat *seat = NULL;
+static void handle_global(void *data, struct wl_registry *registry,
+		uint32_t name, const char *interface, uint32_t version) {
+	if (strcmp(interface, wl_output_interface.name) == 0) {
+		struct wl_output *output = wl_registry_bind(registry, name,
+				&wl_output_interface, version);
+		wl_output_set_user_data(output, (void*)(size_t)name); // assign some ID to the output
+	} else if (strcmp(interface,
+			zwlr_workspace_manager_v1_interface.name) == 0) {
+		workspace_manager = wl_registry_bind(registry, name,
+				&zwlr_workspace_manager_v1_interface, WLR_WORKSPACE_VERSION);
+
+		wl_list_init(&group_list);
+		zwlr_workspace_manager_v1_add_listener(workspace_manager,
+				&workspace_manager_impl, NULL);
+	}
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+		uint32_t name) {
+	// who cares
+}
+
+static const struct wl_registry_listener registry_listener = {
+	.global = handle_global,
+	.global_remove = handle_global_remove,
+};
+
+static struct workspace_v1 *workspace_by_name_or_bail(const char *name) {
+	struct workspace_v1 *workspace;
+	struct group_v1 *group;
+
+	wl_list_for_each(group, &group_list, link) {
+		wl_list_for_each(workspace, &group->workspaces, link) {
+			if (workspace->current.name &&
+					strcmp(workspace->current.name, name) == 0) {
+				return workspace;
+			}
+		}
+	}
+
+	fprintf(stderr, "No workspace with the given name: %s\n", name);
+	exit(EXIT_FAILURE);
+}
+
+int main(int argc, char **argv) {
+	int c;
+	char *focus_name = NULL;
+	bool monitor = false;
+
+	while ((c = getopt(argc, argv, "w:m")) != -1) {
+		switch (c) {
+		case 'w':
+			focus_name = strdup(optarg);
+			break;
+		case 'm':
+			monitor = true;
+			break;
+		}
+	}
+
+	struct wl_display *display = wl_display_connect(NULL);
+	if (display == NULL) {
+		fprintf(stderr, "Failed to create display\n");
+		return EXIT_FAILURE;
+	}
+
+	struct wl_registry *registry = wl_display_get_registry(display);
+	wl_registry_add_listener(registry, &registry_listener, NULL);
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+
+	if (workspace_manager == NULL) {
+		fprintf(stderr, "wlr-workspace not available\n");
+		return EXIT_FAILURE;
+	}
+	wl_display_roundtrip(display); // load workspace groups
+	wl_display_roundtrip(display); // load details
```

Comment:
Why it called twice after 304th line?
As I see here https://wayland.freedesktop.org/docs/html/apb.html
it blocks for all requests

Reply:
The comments say it. On the firsst roundtrip, we get all events for workspaces. We bind all workspaces during this first roundtrip. We also connect our listeners. Then we need a second roundtrip during which we get all details for the workspaces.

Reply:
I see.